segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -0,0 +1,510 @@
+            if (skipStoreCheck || entry.getValue().isPinned()) {
+                log.debug("Skipping loading key from the store key = {}", key);
+            } else {
+                // This check is safe to be outside the lock
+                if (!bufferedTxnData.containsKey(key)) {
+                    loadFromStore(key);
+                }
+            }
+        }
+
+        // Step 2 : Check whether transaction is safe to commit.
+        // This check needs to be atomic, with absolutely no possibility of re-entry
+        synchronized (lock) {
+            for (Map.Entry<String, TransactionData> entry : txnData.entrySet()) {
+                String key = entry.getKey();
+                val transactionData = entry.getValue();
+                Preconditions.checkState(null != transactionData.getKey());
+
+                // See if this entry was modified in this transaction.
+                if (transactionData.getVersion() == txn.getVersion()) {
+                    modifiedKeys.add(key);
+                    transactionData.setPersisted(false);
+                    modifiedValues.add(transactionData);
+                }
+                // make sure none of the keys used in this transaction have changed.
+                TransactionData dataFromBuffer = bufferedTxnData.get(key);
+                if (null != dataFromBuffer) {
+                    if (dataFromBuffer.getVersion() > transactionData.getVersion()) {
+                        throw new StorageMetadataVersionMismatchException(
+                                String.format("Transaction uses stale data. Key version changed key:%s buffer:%s transaction:%s",
+                                        key, dataFromBuffer.getVersion(), txnData.get(key).getVersion()));
+                    }
+
+                    // Pin it if it is already pinned.
+                    transactionData.setPinned(transactionData.isPinned() || dataFromBuffer.isPinned());
+
+                    // Set the database object.
+                    transactionData.setDbObject(dataFromBuffer.getDbObject());
+                }
+            }
+
+            // Step 3: Commit externally.
+            // This operation may call external storage.
+            if (!lazyWrite || (bufferedTxnData.size() > maxEntriesInTxnBuffer)) {
+                log.debug("Persisting all modified keys (except pinned)");
+                val toWriteList =  modifiedValues.stream().filter(entry -> !entry.isPinned()).collect(Collectors.toList());
+                writeAll(toWriteList);
+                log.debug("Done persisting all modified keys");
+
+                // Mark written keys as persisted.
+                for (val writtenData: toWriteList) {
+                    writtenData.setPersisted(true);
+                }
+            }
+
+            // Execute external commit step.
+            try {
+                if (null != txn.getExternalCommitStep()) {
+                    txn.getExternalCommitStep().call();
```

Comment:
You execute quite a number of external APIs, some of them a bit shaky in nature (like this one) while holding your lock. The more APIs you execute externally, and the more async they are in nature, the more likely it is for your application to perform poorly or deadlock. Can you reduce the scope of the encompassing lock?

Reply:
The scope of the lock can not be reduced. However I have plans to take locks per segment instead of global locks.